java/arcs/core/analysis/TypeConstraints.kt

```diff
@@ -0,0 +1,78 @@
+/*
+ * Copyright 2020 Google LLC.
+ *
+ * This code may only be used under the BSD style license found at
+ * http://polymer.github.io/LICENSE.txt
+ *
+ * Code distributed by Google as part of this project is also subject to an additional IP rights
+ * grant found at
+ * http://polymer.github.io/PATENTS.txt
+ */
+
+package arcs.core.analysis
+
+import arcs.core.data.HandleConnectionSpec
+import arcs.core.data.ParticleSpec
+import arcs.core.data.Recipe
+import arcs.core.data.Recipe.Particle
+import arcs.core.data.TypeVariable
+
+/** The possible nodes that can appear in a type constraint. */
+sealed class TypeConstraintNode {
+    /** Representation of a handle connection in a particle. */
+    data class HandleConnection(
+        val particleSpec: ParticleSpec,
+        val connectionSpec: HandleConnectionSpec
+    ) : TypeConstraintNode() {
+        override fun toString() = "${particleSpec.name}.${connectionSpec.name}"
+    }
+
+    /** Representation of a [Recipe.Handle]. */
+    data class Handle(val handle: Recipe.Handle) : TypeConstraintNode() {
+        override fun toString() = "${handle.name}"
+    }
+}
+
+/** A type constraint. The order of the nodes do not matter. i.e., (a, b) is same as (b, a). */
+data class TypeConstraint(val lhs: TypeConstraintNode, val rhs: TypeConstraintNode) {
+    override fun equals(other: Any?): Boolean {
+        if (this === other) return true
+        if (other == null) return false
+        if (this::class != other::class) return false
+
+        other as TypeConstraint
```

Comment:
What does this do? I would expect this to drop the (now cast) value of other.

Reply:
Yep, it  casts `other` to `TypeConstraint`. After this point, we can access the fields of TypeConstraint through `other` directly.